pylint/checkers/variables.py

```diff
@@ -2282,20 +2282,37 @@ def _is_variable_violation(
         elif isinstance(defframe, nodes.ClassDef) and isinstance(
             frame, nodes.FunctionDef
         ):
-            # Special rule for function return annotations,
-            # using a name defined earlier in the class containing the function.
-            if node is frame.returns and defframe.parent_of(frame.returns):
-                annotation_return = True
-                if frame.returns.name in defframe.locals:
-                    definition = defframe.locals[node.name][0]
-                    if definition.lineno is None or definition.lineno < frame.lineno:
-                        # Detect class assignments with a name defined earlier in the
-                        # class. In this case, no warning should be raised.
-                        maybe_before_assign = False
+            # Special rules for function return annotations.
+            if node is frame.returns:
+                # Using a name defined earlier in the class containing the function.
+                if defframe.parent_of(frame.returns):
+                    annotation_return = True
+                    if frame.returns.name in defframe.locals:
+                        definition = defframe.locals[node.name][0]
+                        if (
+                            definition.lineno is None
+                            or definition.lineno < frame.lineno
+                        ):
+                            # Detect class assignments with a name defined earlier in the
+                            # class. In this case, no warning should be raised.
+                            maybe_before_assign = False
+                        else:
+                            maybe_before_assign = True
                     else:
                         maybe_before_assign = True
-                else:
-                    maybe_before_assign = True
+                # Using a name defined in the module if this is a nested function.
+                elif (
+                    (defframe_parent := next(defframe.node_ancestors()))
+                    and isinstance(defframe_parent, nodes.Module)
+                    and (frame_ancestors := tuple(frame.node_ancestors()))
+                    and any(
+                        isinstance(ancestor, nodes.FunctionDef)
+                        for ancestor in frame_ancestors
+                    )
+                    and frame_ancestors[-1] is defframe_parent
```

Comment:
I took around 5 minute to understand this and I'm still unsure if I do. (It's not a very useful comment, but this part of the code is starting to frighten me :) )